ui/app/common/daclient/jsonrpc.js

```diff
@@ -20,158 +20,182 @@
-            return socket.send(msg);
-          });
-        }
+  /** 
+   * This provider can be enhanced with interceptors:
+   *
+   * module.config(['jsonrpcProvider', function(jsonrpcProvider) {
+   *   jsonrpcProvider.interceptors.push(function(){
+   *     return {
+   *       requestStarted: function(requestFinishedPromise, cfpLoadingBar) { ..custom logic }
+   *     }
+   *   });
+   * }]);
+   */
+  module.provider('jsonrpc', function() {
+
+    this.interceptors = [];
+
+    this.$get = [
+      '$log',
+      '$q',
+      '$websocket',
+      'cfpLoadingBar',
+      function($log, $q, $websocket, cfpLoadingBar) {
+
+        var that = this;
+
+        var JSON_RPC_VERSION = '2.0';
+        var ID_PREFIX = 'request_';
+
+        function wsClient(url) {
+          var instance = {};
+
+          /*
+           * Instance private fields
+           */
+
+          var socket;
+          var deferrals = {}; // Map of id => deferred
+
+          /*
+           * Instance private methods
+           */
+          function sendOnSocket(msg) {
+            return instance.connect().then(function() {
+              return socket.send(msg);
+            });
+          }
 
-        function makeRpcRequest(request) {
-          var deferred = $q.defer();
+          function makeRpcRequest(request) {
+            var deferred = $q.defer();
 
-          request.id = _.uniqueId(ID_PREFIX);
+            request.id = _.uniqueId(ID_PREFIX);
 
-          // As rpc invocations are asynchronous and request -> response may
-          // interleave with other requests we store the id => deferred entry
-          // in the map so we can resolve / reject  the correct promise
-          // when we receive the rpc response.
-          deferrals[request.id] = deferred;
+            // As rpc invocations are asynchronous and request -> response may
+            // interleave with other requests we store the id => deferred entry
+            // in the map so we can resolve / reject  the correct promise
+            // when we receive the rpc response.
+            deferrals[request.id] = deferred;
 
-          sendOnSocket(request).catch(function () {
-            deferred.reject({
-              code: 0,
-              message: 'Error connecting to server'
+            sendOnSocket(request).catch(function () {
+              deferred.reject({
+                code: 0,
+                message: 'Error connecting to server'
+              });
             });
-          });
 
-          $log.debug('Making RPC request: ' + JSON.stringify(request, null, 2));
+            $log.debug('Making RPC request: ' + JSON.stringify(request, null, 2));
 
-          return deferred.promise;
-        }
-
-        function handleRpcResponse(response) {
-          $log.debug('Received RPC response: ' + JSON.stringify(response, null, 2));
+            that.interceptors.forEach(function(interceptor) {
+              interceptor().requestStarted(deferred.promise, cfpLoadingBar);
```

Comment:
I wouldn't inject the cfpLoading bar here, this tightly couples the interceptor logic. I would just inject the promise for each interceptor and access the loading bar where you create the interceptor itself. I'll write a general comment with how I would achieve this.